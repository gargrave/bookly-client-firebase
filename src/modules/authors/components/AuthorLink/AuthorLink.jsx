// @flow
import React from 'react';
import { Link } from 'react-router-dom';
import { object } from 'prop-types';

import type { Author } from '../../flowtypes';

import { localUrls } from '../../../../globals/urls';

import styles from './AuthorLink.css';

type Props = {
  author: Author,
  history?: Object,
};

const AuthorLink = ({
  author,
}: Props) => {
  const authorId = author.id || '';
  return (
    <p className={styles.authorLink}>
      by&nbsp;
      <Link to={`${localUrls.authorsList}/${authorId}`}>
        {author.firstName} {author.lastName}
      </Link>
    </p>
  );
};

AuthorLink.propTypes = {
  author: object.isRequired,
  history: object,
};

export default AuthorLink;
