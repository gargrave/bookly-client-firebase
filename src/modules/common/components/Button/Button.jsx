// @flow
import React from 'react'
import { bool, func, oneOf, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import styles from './Button.css'

export const POSITIONS = ['left', 'right']

export const TYPES = [
  'success',
  'secondary',
  'info',
  'warning',
  'danger',
  'light',
  'dark',
]

type Props = {
  canSubmit?: boolean,
  disabled?: boolean,
  onClick: Function,
  position?: string,
  text: string,
  type?: string,
}

const Button = ({
  canSubmit,
  disabled,
  onClick,
  position = '',
  text,
  type = '',
}: Props) => (
  <button
    className={buildClass(
      styles.button,
      { [styles[position]]: !!position },
      'button',
      { [`button-${type}`]: !!type },
    )}
    disabled={disabled || false}
    onClick={onClick}
    type={canSubmit ? 'submit' : 'button'}
  >
    {text}
  </button>
)

Button.propTypes = {
  canSubmit: bool,
  disabled: bool,
  onClick: func.isRequired,
  position: oneOf(POSITIONS),
  text: string.isRequired,
  type: oneOf(TYPES),
}

export default Button
