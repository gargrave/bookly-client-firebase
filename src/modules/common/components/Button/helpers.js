// @flow
import { buildClasses } from '../../../../globals/utils/cssHelpers';

import {
  POSITIONS,
  TYPES,
} from './constants';

const buttonClass = (type: string) => {
  if (TYPES.includes(type)) {
    return `button-${type}`;
  }
  return '';
};

const positionClass = (position: string) => {
  if (POSITIONS.includes(position)) {
    return `button--${position}`;
  }
  return '';
};

export const buildClassList = (
  type: string,
  position: string,
  classes: string,
) => {
  return buildClasses(
    ['button', positionClass(position)],
    [buttonClass(type), ...classes.split(' ')],
  );
};
