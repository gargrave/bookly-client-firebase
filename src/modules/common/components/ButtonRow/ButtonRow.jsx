// @flow
import React from 'react'
import { array, object, oneOfType } from 'prop-types'

import styles from './ButtonRow.css'

type Props = {
  children?: any,
}

const ButtonRow = ({ children }: Props) => {
  return <div className={styles.buttonRow}>{children}</div>
}

ButtonRow.propTypes = {
  children: oneOfType([array, object]),
}

export default ButtonRow
