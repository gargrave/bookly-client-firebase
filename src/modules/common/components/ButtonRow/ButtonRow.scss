.buttonRow {
  display: flex;
  justify-content: space-between;
}
