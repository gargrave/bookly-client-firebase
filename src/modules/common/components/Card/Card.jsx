// @flow
import React from 'react'
import { any, bool, func } from 'prop-types'
import styled from 'react-emotion'

import { buildClass } from '../../../../utils/cssHelpers'

import CardHeader from './CardHeader'
import CardSpacer from './CardSpacer'
import CardTextLine from './CardTextLine'
import CardTextList from './CardTextList'

import { colors, shadows } from '../../../../styles/'

type Props = {
  children?: any,
  hoverable?: boolean,
  onClick?: Function,
}

const Styled = styled('div')`
  ${shadows.medium};

  background-color: ${colors.cardBg};
  border: solid 1px ${colors.cardBorder};
  border-radius: 5px;
  margin: 8px auto;
  max-width: 550px;
  padding: 10px 20px;
  text-align: center;

  &.hoverable:hover {
    background: ${colors.cardHoveredBg};
    cursor: pointer;
  }
`

const Card = ({ children, hoverable = false, onClick }: Props) => {
  return (
    <Styled className={buildClass({ hoverable })} onClick={onClick}>
      {children}
    </Styled>
  )
}

Card.Header = CardHeader
Card.Spacer = CardSpacer
Card.TextLine = CardTextLine
Card.TextList = CardTextList

Card.propTypes = {
  children: any,
  hoverable: bool,
  onClick: func,
}

export default Card
