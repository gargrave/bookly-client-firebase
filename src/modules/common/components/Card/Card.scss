@import '../../../../styles/_shadows';

$textColor: #6e6969;
$titleColor: #5d5656;

.card {
  @include boxShadowMed;
  background-color: #f9f9f9;
  border: solid 1px #bbb;
  border-radius: 5px;
  margin: 8px auto;
  max-width: 550px;
  padding: 10px 20px;
}

.hoverableCard:hover {
  background: #e4effc;
  cursor: pointer;
}

.header,
.title,
.text {
  margin: 0;
}

.header,
.title {
  color: $titleColor;
  font-weight: bold;
}

.header {
  font-size: 1.4rem;
}

.detailCard {
  padding: 15px;

  .title {
    font-size: 1.3rem;
    line-height: 2.5rem;
  }

  .title + .cardText {
    font-size: 1.1rem;
  }

  .cardText + .cardText {
    margin-top: 8px;
  }
}

.listDetailCard {
  padding: 10px
}
