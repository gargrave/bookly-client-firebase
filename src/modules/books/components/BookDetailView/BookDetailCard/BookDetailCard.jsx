// @flow
import React from 'react';
import { func, instanceOf, oneOfType, shape, string } from 'prop-types';
import { format } from 'date-fns';

import type { Book } from '../../../flowtypes';

import AuthorLink from '../../../../authors/components/AuthorLink/AuthorLink';
import Button from '../../../../common/components/Button/Button';
import ButtonRow from '../../../../common/components/ButtonRow/ButtonRow';
import Card from '../../../../common/components/Card/Card';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
};

const bookDatesTextList = (book: Book) => {
  const {
    created,
    updated,
  } = book;

  return [
    { title: 'Added', value: format(created, 'MMM. DD, YYYY, HH:mm:ss') },
    { title: 'Updated', value: format(updated, 'MMM. DD, YYYY, HH:mm:ss') },
  ];
};

const BookDetailCard = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
}: Props) => {
  const {
    title,
  } = book;

  return (
    <Card
      header={title}>
      <AuthorLink
        author={book.author} />

      <Card.TextList textList={bookDatesTextList(book)} />
      <Card.Spacer size='large' />

      <ButtonRow>
        <Button
          onClick={onEditClick}
          position="left"
          text="Edit"
          type="info" />
        <Button
          onClick={onDeleteClick}
          text="Delete"
          type="danger" />
        <Button
          onClick={onBackClick}
          position="right"
          text="Back"
          type="light" />
      </ButtonRow>
    </Card>
  );
};

BookDetailCard.propTypes = {
  book: shape({
    created: oneOfType([
      instanceOf(Date),
      string,
    ]),
    updated: oneOfType([
      instanceOf(Date),
      string,
    ]),
    author: shape({
      firstName: string,
      lastName: string,
    }),
    title: string,
  }).isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
};

export default BookDetailCard;
