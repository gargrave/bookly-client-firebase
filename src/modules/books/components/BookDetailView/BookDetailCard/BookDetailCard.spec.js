import { format } from 'date-fns'

import { bookMocks } from '../../../../../globals/mocks/'
import { ComponentBuilder } from '../../../../../utils/testHelpers'

import AuthorLink from '../../../../authors/components/AuthorLink/AuthorLink'
import Button from '../../../../common/components/Button/Button'
import ButtonRow from '../../../../common/components/ButtonRow/ButtonRow'
import Card from '../../../../common/components/Card/Card'

import BookDetailCard from './BookDetailCard'

const book = { ...bookMocks[0] }
const defaultProps = {
  book,
  onBackClick: jest.fn(),
  onDeleteClick: jest.fn(),
  onEditClick: jest.fn(),
}

const builder = new ComponentBuilder(BookDetailCard, defaultProps)

const dateFormat = 'MMM. DD, YYYY, HH:mm:ss'

describe('BookDetailCard', () => {
  let component

  it('matches the snapshot', () => {
    component = builder.shallowGetComponent()
    expect(component).toMatchSnapshot()
  })

  describe('rendering', () => {
    it('renders correctly', () => {
      component = builder.shallowGetComponent()
      expect(component.find(AuthorLink)).toHaveLength(1)
      expect(component.find(Card)).toHaveLength(1)
      expect(component.find(Card.TextList)).toHaveLength(1)
      expect(component.find(ButtonRow)).toHaveLength(1)
      expect(component.find(Button)).toHaveLength(3)
    })

    it('renders the book title in the Card header', () => {
      component = builder.shallowGetComponent()
      const card = component.find(Card)
      expect(card.props().header).toBe(book.title)
    })

    it('renders the correct text lines', () => {
      component = builder.shallowGetComponent()
      const textList = component.find(Card.TextList).props().textList
      expect(textList.length).toBe(3)

      const sortByText = textList.find(line => line.title === 'Sort by')
      expect(sortByText).toBeDefined()
      expect(sortByText.value).toBe(book.sortBy)

      const addedText = textList.find(line => line.title === 'Added')
      expect(addedText).toBeDefined()
      expect(addedText.value).toBe(format(book.created, dateFormat))

      const updatedText = textList.find(line => line.title === 'Added')
      expect(updatedText).toBeDefined()
      expect(updatedText.value).toBe(format(book.created, dateFormat))
    })
  })
})
