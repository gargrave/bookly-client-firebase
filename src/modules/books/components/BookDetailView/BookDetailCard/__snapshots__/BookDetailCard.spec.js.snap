// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetailCard matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookDetailCard
    book={
      Object {
        "author": Object {
          "bookCount": 1,
          "created": 2017-07-01T01:16:43.881Z,
          "firstName": "Aldous",
          "id": "oiw223h3qndlksadffl3ksdf",
          "lastName": "Huxley",
          "updated": 2017-07-01T01:16:43.881Z,
        },
        "created": 2018-01-14T01:36:26.220Z,
        "id": "sv3hmbefpc95MRXQsTHr",
        "title": "Brave New World",
        "updated": 2018-01-14T01:36:26.220Z,
      }
    }
    onBackClick={[Function]}
    onDeleteClick={[Function]}
    onEditClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <AuthorLink
          author={
            Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            }
          }
        />,
        <CardTextList
          textList={
            Array [
              Object {
                "title": "Sort by",
                "value": undefined,
              },
              Object {
                "title": "Added",
                "value": "Jan. 13, 2018, 17:36:26",
              },
              Object {
                "title": "Updated",
                "value": "Jan. 13, 2018, 17:36:26",
              },
            ]
          }
        />,
        <CardSpacer
          size="medium"
        />,
        <ButtonRow>
          <Button
            onClick={[Function]}
            position="left"
            text="Edit"
            type="info"
          />
          <Button
            onClick={[Function]}
            text="Delete"
            type="danger"
          />
          <Button
            onClick={[Function]}
            position="right"
            text="Back"
            type="light"
          />
        </ButtonRow>,
      ],
      "header": "Brave New World",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "author": Object {
            "bookCount": 1,
            "created": 2017-07-01T01:16:43.881Z,
            "firstName": "Aldous",
            "id": "oiw223h3qndlksadffl3ksdf",
            "lastName": "Huxley",
            "updated": 2017-07-01T01:16:43.881Z,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "textList": Array [
            Object {
              "title": "Sort by",
              "value": undefined,
            },
            Object {
              "title": "Added",
              "value": "Jan. 13, 2018, 17:36:26",
            },
            Object {
              "title": "Updated",
              "value": "Jan. 13, 2018, 17:36:26",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "size": "medium",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Button
              onClick={[Function]}
              position="left"
              text="Edit"
              type="info"
            />,
            <Button
              onClick={[Function]}
              text="Delete"
              type="danger"
            />,
            <Button
              onClick={[Function]}
              position="right"
              text="Back"
              type="light"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "position": "left",
              "text": "Edit",
              "type": "info",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "text": "Delete",
              "type": "danger",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "position": "right",
              "text": "Back",
              "type": "light",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <AuthorLink
            author={
              Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              }
            }
          />,
          <CardTextList
            textList={
              Array [
                Object {
                  "title": "Sort by",
                  "value": undefined,
                },
                Object {
                  "title": "Added",
                  "value": "Jan. 13, 2018, 17:36:26",
                },
                Object {
                  "title": "Updated",
                  "value": "Jan. 13, 2018, 17:36:26",
                },
              ]
            }
          />,
          <CardSpacer
            size="medium"
          />,
          <ButtonRow>
            <Button
              onClick={[Function]}
              position="left"
              text="Edit"
              type="info"
            />
            <Button
              onClick={[Function]}
              text="Delete"
              type="danger"
            />
            <Button
              onClick={[Function]}
              position="right"
              text="Back"
              type="light"
            />
          </ButtonRow>,
        ],
        "header": "Brave New World",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "author": Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "textList": Array [
              Object {
                "title": "Sort by",
                "value": undefined,
              },
              Object {
                "title": "Added",
                "value": "Jan. 13, 2018, 17:36:26",
              },
              Object {
                "title": "Updated",
                "value": "Jan. 13, 2018, 17:36:26",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "size": "medium",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Button
                onClick={[Function]}
                position="left"
                text="Edit"
                type="info"
              />,
              <Button
                onClick={[Function]}
                text="Delete"
                type="danger"
              />,
              <Button
                onClick={[Function]}
                position="right"
                text="Back"
                type="light"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "position": "left",
                "text": "Edit",
                "type": "info",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "text": "Delete",
                "type": "danger",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "position": "right",
                "text": "Back",
                "type": "light",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
